File: 0073-set-matrix-zeroes/0073-set-matrix-zeroes.c
unsigned char row[25];
unsigned char col[25];

int test(char c,int n) {
    if(c=='r') {
        return row[n>>3]&(1<<(n&7));
    }
    else {
        return col[n>>3]&(1<<(n&7));
    }
}

void set(char c,int n) {
    if(c=='r') {
        row[n>>3]|=1<<(n&7);
    }
    else {
        col[n>>3]|=1<<(n&7);
    }
}


void setZeroes(int** matrix, int matrixSize, int* matrixColSize) {
    int i,j;
    int m=matrixSize;
    int n=matrixColSize[0];
    memset(row,0,sizeof(row));
    memset(col,0,sizeof(col));
    for(i=0;i<m;i++) {
        for(j=0;j<n;j++) {
            if(matrix[i][j]==0) {
                set('r',i);
                set('c',j);
            }
        }
    }
    for(i=0;i<m;i++) {
        if(test('r',i)) {
            for(j=0;j<n;j++) {
                matrix[i][j]=0;
            }
        }
    }
    for(i=0;i<n;i++) {
        if(test('c',i)) {
            for(j=0;j<m;j++) {
                matrix[j][i]=0;
            }
        }
    }
}

File: 0073-set-matrix-zeroes/0073-set-matrix-zeroes.py
class Solution:
    def setZeroes(self, matrix: List[List[int]]) -> None:
        m = len(matrix)
        if m == 0: return
        n = len(matrix[0])

        first_row = matrix[0]
        first_row_zero = any(x == 0 for x in first_row)
        first_col_zero = any(matrix[i][0] == 0 for i in range(m))

        for i in range(1, m):
            row = matrix[i]
            for j in range(1, n):
                if row[j] == 0:
                    row[0] = 0
                    first_row[j] = 0

        for i in range(1, m):
            row = matrix[i]
            mark = row[0]
            for j in range(1, n):
                if mark == 0 or first_row[j] == 0:
                    row[j] = 0

        if first_row_zero:
            for j in range(n):
                first_row[j] = 0

        if first_col_zero:
            for i in range(m):
                matrix[i][0] = 0